### ISSUE TITLE
Lint to warn on use of `#[deny(lint)]` in code

### ISSUE THREAD

**Issue by epage** [rust-lang/rust-clippy#13963]
### What it does

Warn on use of `#[deny(lint)]` / `#![deny(lint)]`

### Advantage

- Avoid builds breaking due to a new Rust toolchain changing the behavior of a lint
- Give developers flexibility to be sloppy during development, resolving out the lints before merging (generally being enforced in CI)

### Drawbacks

- It becomes awkward to `allow` this warning.  A parent scope can do it.  Unsure if you could in the same scope with `!`.  `Cargo.toml` `[lints]` could also

### Example

```rust
#![deny(elided-lifetimes-in-paths)]

pub struct Foo<'a> {}

fn foo(f: Foo) {}
```

Could be written as:

```rust
#![warn(elided-lifetimes-in-paths)]

pub struct Foo<'a> {}

fn foo(f: Foo) {}
```
with the suggestion to turn warnings into errors where relevant (e.g. CI)

**Comment by y21** [comment]
IMO there are a lot of perfectly valid and good use cases for local denys, e.g. `#[deny(clippy::unwrap_used)]` on a `GlobalAlloc` impl where unwinding would be undefined behavior, or `#[deny(clippy::missing_trait_methods)]` on a trait impl for a wrapper type that must delegate all method calls to the wrapped type. In those cases it seems right to outright deny builds and not allow that at all even during development because it leads to wrong behavior.
I don't think those use cases should have to allow the lint.

**Comment by epage** [comment]
How is that different than `warn` and then denying?

The only reason I could see for putting those in is to instead `forbid`

**Comment by y21** [comment]
Can you explain what you mean by "and then denying"? Like, `#[warn(clippy::unwrap_used)]` and then `-Dclippy::unwrap_used` on the command line? That globally denies the restriction lint for the whole crate when it's only relevant for one specific impl. Or `-Dwarnings`? That will make all warnings into errors, not that one particular warning.

I have my editor set up to run clippy on save, so `#[deny]` would be used in that situation to have clippy treat that specific instance more like a compiler error and so that it stands out from e.g. `dead_code` which can be mostly ignored during prototyping/quick edit-run cycles.

**Comment by epage** [comment]
So if I'm understanding correctly, these cases are important enough that you want immediate, blocking feedback on, rather than seeing it in the list of warnings and deciding to act on it?

Maybe I'm biased by my own workflows or the types of projects I work on but I've never needed to do that and I suspect it would be limited if I did step into areas you mentioned, like `GlobalAlloc`, that this doesn't take away from the value of this proposed lint but either suggests that people may need to occasionally `allow` .  It could be `allow` by default but I think that would remove most of the benefit as this would be helping people discover a common, happy path for development rather than having to re-discover that `deny` shouldn't generally be used.

**Comment by hanna-kruppe** [comment]
I think this would be a good lint to have and to *warn by default*. For every expert user who strategically applies `#[deny(specific_lint)]`, I've seen dozens of projects that set up a big future compatibility footgun for themselves with something like crate-level `#![deny(warnings)]`. For third party dependencies, wider repercussions are prevented by `--cap-lints`, but it still causes problems and friction sometimes, such as:

- People mistakenly treating new warnings as "breaking changes in Rust" because someone's `deny(warnings)` + "latest stable" CI workflow broke (see https://github.com/actions/runner-images/pull/13044#issuecomment-3321290737 for a recent high profile example)
- People working on the same project using different toolchain versions than what's tested in CI (more than once I've wanted to prepare a quick PR for someone else's project, and saw a bunch of new clippy warnings in the project's preferred clippy configuration).
- Certain lints firing in downstream code even though the root cause is upstream (e.g., https://github.com/rust-lang/rust/issues/149845)
- Rough edges around configuring warnings for different rustc drivers like rustdoc (e.g., https://github.com/rust-lang/rust/issues/67533)

On balance, less experienced Rust users are best served from being guided away from `#[deny(...)]` and towards CI practices. I appreciate that there are different workflows which knowingly accept the downsides of always-on `deny`, but clippy (unlike rustc) can and often does ask people to manually turn off lints they disagree with.

Another consideration is consistency with Cargo, which may lint against `deny` in the Cargo.toml `[lints]` table in the future: https://github.com/rust-lang/cargo/issues/16667

Unfortunately, even Clippy's own README still heavily suggests `#![deny(...)]`. Manually passing CLI flags to `cargo clippy` is the main alternative presented, `RUSTFLAGS` is not even mentioned. Per `git blame` all those parts of the README seem fairly old, so maybe it's just due for modernization.
